### DISCUSSION TITLE
Trac-IK plugin

### DISCUSSION THREAD

**Discussion by rjoomen** [tesseract-robotics/tesseract#918] [Ideas]
I've created a tesseract_kinematics plugin for the [Trac-IK inverse kinematics solver](https://traclabs.com/projects/trac-ik/). Would you be interested in including this in the main repo?

Problem is that it currently depends on a [non-ROS-agnostic ROS2 fork](https://github.com/aprotyas/trac_ik). The main repo now [also is updated for ROS2](https://bitbucket.org/traclabs/trac_ik/src/rolling-devel/), but is even less ROS-agnostic (needs a node as parameter).

If we'd want to include this, probably forking it under tesseract_robotics and making the trac_ik_lib subproject ROS-agnostic would be the way to go. Afterwards we should try to get this restructured version PR'd to the original repo.

**Comment by Levi-Armstrong** [comment]
I am interested in getting track ik integrated. I would post the question on track ik to see if they would be interested in making track_ik_lib ROS-agnostic. If not, it looks like the track_ik_lib is stable so we could just copy the contents into tesseract_kinematics along with your plugin and make it an optional build.

**Comment by rjoomen** [comment]
I've added the plugin as a [new project](https://github.com/tesseract-robotics/tesseract_trac_ik), pending the creation of a ROS-agnostic Trac-IK fork. Is this OK with you?

**Reply by gavanderhoorn** [reply]
I would suggest just use the [upstream `rolling-devel` version](https://bitbucket.org/traclabs/trac_ik/src/rolling-devel/). The fork you mention is outdated and no longer maintained/developed.

As to a ROS-agnostic version: there are already quite some forks which add that ability. See [HIRO-group/trac_ik](https://github.com/HIRO-group/trac_ik) fi.

**Reply by rjoomen** [reply]
Upstream rolling-devel requires [a rclcpp::Node as an argument](https://bitbucket.org/traclabs/trac_ik/src/5be8a18c7dbe084366e8e61605f37c065151189c/trac_ik_lib/src/trac_ik.cpp#lines-42), which Tesseract cannot provide. The fork I now use [does not](https://github.com/aprotyas/trac_ik/blob/ros2/trac_ik_lib/src/trac_ik.cpp#L48).

**Comment by rjoomen** [comment]
My PR to Trac_IK has been accepted: https://bitbucket.org/traclabs/trac_ik/pull-requests/40

`tesseract_trac_ik` now works with the official Trac_IK repo, instead of a fork. Trac_IK is still ROS-dependent, but for anyone using Tesseract with ROS, this is new IK plugin is a viable option.
